File: src/ga/cross/SinglePointCrossover.js

'use strict';

const utils = require('../../utils')
  , Crossover = require('./Crossover');

class SinglePointCrossover extends Crossover {

  _crossover(pr1, pr2) {
    // Last point of crossover is last but one
    const pos = utils.math.randomInt(0, pr1.length - 2);

    // Exchange params of chromosomes from 0 to `pos`
    for (let i = 0; i <= pos; i += 1) {
      let tmp = pr1[i];
      pr1[i] = pr2[i];
      pr2[i] = tmp;
    }
  }

}

module.exports = SinglePointCrossover;

File: src/core/main/java/workscheduler/genetics/crossover/SinglePointCrossover.java
package workscheduler.genetics.crossover;

import workscheduler.genetics.Chromosome;
import workscheduler.utils.MathUtils;
import workscheduler.utils.Pair;

public final class SinglePointCrossover extends AbstractCrossover {
  @Override
  protected void implementation(final Pair<Chromosome> parents) {
    // Last point of crossover is last but one
    int pos = MathUtils.randomInt(0, parents.one.getLength() - 2);

    // Exchange params of chromosomes from 0 to `pos`
    for (int i = 0; i <= pos; i += 1) {
      boolean tmp = parents.one.getGene(i);
      parents.one.setGene(i, parents.two.getGene(i));
      parents.two.setGene(i, tmp);
    }
  }
}
